### DISCUSSION TITLE
Recovery options for breaking configuration changes

### DISCUSSION THREAD

**Discussion by jesseschmieg** [keycloak/keycloak#12737] [Q&A]
Hello

As I've been exploring Keycloak, I've noticed a few different actions that can be unrecoverable.

For example, removing the realm client-level (`<realm name>-realm`) roles from the master `admin` role. After doing this, I'm unable to re-add the roles and the realm is in a state that it cannot be managed, deleted, etc.

Another example was adding a realm through [keycloak-benchmark](https://github.com/keycloak/keycloak-benchmark) and, oddly enough, the `<realm name>-realm` client was not added and all permission checks from the master realm began throwing NPEs.

For many of these cases, yes, it would be best to avoid allowing them to happen at all. And they could certainly be recovered through fine database operations and recreating the configurations. However, I'm wondering if there are any advisable solutions for otherwise recovering or having some "factory-reset" options for these basic realm configurations.

Thanks in advance for any advice

**Comment by OmgImAlexis** [comment]
@jesseschmieg did you find an answer to this?

**Reply by souravs17031999** [reply]
I believe only current solution is to take a backup before any critical operations and restore if things go wrong.
